Add partial admission control to Thrift frontend

psarna opened the pull request:
This pull request adds partial admission control to Thrift frontend. The solution is partial mostly because the Thrift layer, aside from allowing Thrift messages, may also be used as a base protocol for CQL messages. Coupling admission control to this one is a little bit more complicated due to how the layer currently works - a Thrift handler, created once per connection, keeps a local `query_state` instance for the occasion of handling CQL requests. However, `query_state` should be kept per query, not per connection, so adding admission control to this aspect of the frontend is left for later.

Finally, the way service permits are passed from the server, via the handler factory, handler and then to queries is hacky. I haven't figured out how to force Thrift to pass custom context per query, so the way it works now is by relying on the fact that the server does not yield (in Seastar sense) between having read the request and launching the proper handler. Due to that, it's possible to just store the service permit in the server itself, pass the reference (address) to it down to the handler, and then read it back from the handling code and claim ownership of it. It works, but if anyone has a better idea, please share.

Refs #4826

psarna commented:
Sample grafana output from the newly added metrics:
![thrift_admission](https://user-images.githubusercontent.com/10433047/111651329-69bb5c80-8806-11eb-9869-b41173d57338.png)

psarna commented:
Without https://github.com/scylladb/scylla/pull/8318 applied, unrelated crashes might occur during testing.

psarna commented:
v2:
 - added support for the `max_concurrent_requests_per_shard` config value, which is now respected by Thrift frontend

psarna commented:
The support is still partial and does not include CQL-on-Thrift, but it's nonetheless ready for review.

nyh commented on thrift/server.cc:
```diff
@@ -148,18 +149,17 @@ future<>
+        _processor_promise = promise<>();
+    };
+    _processor->process(complete, _in_proto, _out_proto);
+    co_return co_await ret;
```
co_return co_await is slightly less efficient than what we had previously (it needs one more continuation), but I am guessing this deliberate and you'll need it later to keep the permit alive.
In any case, I guess the cost of the extra continuation is tiny, and perhaps even offset by other differences beween the coroutine and future version - so I think this change is fine.

nyh commented on thrift/server.cc:
```diff
@@ -151,15 +154,21 @@ thrift_server::connection::process_one_request() {
-    auto ret = _processor_promise.get_future();
+    auto ret = _processor_promise.get_future().handle_exception([&server = _server] (const std::exception_ptr&) {
+        server._requests_serving--;
+    });
```
Nitpick: are the changes here to this function related to this patch? Maybe they should be folded into the previous patch?

nyh commented on main.cc:
```diff
@@ -1282,7 +1282,7 @@ int main(int ac, char** av) {
             });
 
-            ::thrift_controller thrift_ctl(db, auth_service, qp);
+            ::thrift_controller thrift_ctl(db, auth_service, qp, service_memory_limiter);
```
Commenting on this line just because I can't comment on the commit message :-(
In the commit message you say that "This mechanism heavily relies on the fact that there are zero preemption points between overwriting the service permit and reading it back by the handler.". However, I couldn't figure out where in the actual code this assumption is made. Could we have a FIXME in the actual code that assumes this?

psarna replied:
This assumption is a little convoluted, because it's spread over 4 separate entities - the Thrift server, the handler factory, the handler, and then the code which handles a single request; but I'll try to put a comment somewhere - it won't really be a FIXME, because I have no idea how it should be "fixed". The idea is that once a connection reads a request from its socket, it creates a service permit object which is used for reserving resources. Then, this connection puts this service permit in a special place in the **server** object - this one is tricky, because multiple connections share a single server - and then the code which handles a request gets the permit back from this special place in the server, which is global for all connections. And the assumption is that there is no preemption point between putting the service permit in this special place and retrieving it in the code which runs the request. Otherwise we'd have races, because this address is effectively shared between connections. And all that is just because the Thrift API doesn't offer any easy way of passing custom context to a single request handler, or at least I wasn't able to find such an API. A potentially safer alternative to that would be if the server kept a map of <connection, service_permit>, which would keep a separate place to pass the permit per each connection separately. That's additional overhead though.

nyh replied:
So even if not a map, maybe at least check if the server already has a permit set - and if it is, log an internal error? This will allow us to find such a bug if it ever happens because of future code changes. It will also give you a good place to put a FIXME (in front of this logging code) and the FIXME can include this explanation which you wrote here because obviously in a year nobody will remember why you made this assumption.

psarna replied:
Sure, but I'd limit this error to debug/trace, because I'd actually expect it to fire for cql-over-thrift requests (of which I wasn't even aware until I touched this part of our code base), because this series doesn't address this part of Thrift yet.

nyh commented on thrift/server.cc:
```diff
@@ -165,7 +165,11 @@ thrift_server::connection::process_one_request() {
     size_t mem_estimate = 8000 + 2 * _input->available_read();
-    auto units = co_await get_units(_server._memory_available, mem_estimate);
+    auto fut = get_units(_server._memory_available, mem_estimate);
+    if (_server._memory_available.waiters()) {
```
Nitpick: Does waiters() mean that *we* will need to wait?
Wouldn't fut.available() be a more direct way to check if we need to wait or not?

psarna replied:
Yes, it's for checking if we will wait. I did it like that just because the original code in transport/server.cc is almost identical - and there's some value in keeping these two similar, since people familiar with the cql server code would immediately understand what's going on in here.

nyh commented on thrift/server.cc:
```diff
@@ -352,6 +352,20 @@ thrift_stats::thrift_stats(thrift_server& server) {
         sm::make_derive("served", [&server] { return server.requests_served(); },
                         sm::description("Rate of serving Thrift requests.")),
+        sm::make_gauge("serving", [&server] { return server.requests_serving(); },
+                        sm::description("Number of requests being currently served.")),
```
In one counter below you explicitly mentioned that this counter is for the "Thrift transport component", but in this counter and a few others you don't. I think we should try to be more consistent. I would add the mention of Thrift to all the counters - even if the metric's name also includes the word "thrift".

psarna replied:
Sure, makes sense, will add in v2.

nyh commented on thrift/server.cc:
```diff
@@ -362,6 +373,9 @@ thrift_stats::thrift_stats(thrift_server& server) {
                             seastar::format("Holds an incrementing counter with the requests that ever blocked due to reaching the memory quota limit ({}B). "
                                             "The first derivative of this value shows how often we block due to memory exhaustion in the \"Thrift transport\" component.", server.max_request_size()))),
+        sm::make_derive("requests_shed", [&server] { return server.requests_shed(); },
+                        sm::description("Holds an incrementing counter with the requests that were shed due to overload (threshold configured via max_concurrent_requests_per_shard). "
```
The word "overload" is too overloaded :-) Instead of "due to overload" I think it will be clearer to say "due to a configured limit on the number of concurrent Thrift requests".
By the way, I put the word "Thirft" in this suggestion because, if I understand correctly (but please correct me if I'm wrong), the concurrency of different types of requests (CQL, Thrift, Alternator) is limited separately

psarna replied:
Alright, but in that case we should also reword the corresponding metrics in CQL (in a separate follow-up patch) - so that the two can be clearly distinguished as doing almost the same thing.

nyh reviewed (approved):
In general I think this series is good, but I left a few minor comments that would improve it further.
I'm not at all familiar with the Thrift code, so I'd appreciate it if somebody who is familiar with that code will also review.

psarna commented:
v2:
 - updated metrics descriptions to mention Thrift
 - added support for max_concurrent_requests_per_shard config variable
 - added a comment explaining how service permits are passed from the Thrift server to the handlers

nyh commented:
Thanks. I left a suggestion above on how you can recognize that you have a problem that two connections tried to set a permit and log an internal error - so if we ever have the kind of races that you hope we don't, we'll know about them.

psarna commented:
v3:
 - added a debug message if the server tries to overwrite a service permit with units held (which generally should not happen, but also *could* happen right now due to the partial nature of this series)

nyh commented on thrift/server.cc:
```diff
@@ -158,8 +161,23 @@ thrift_server::connection::process_one_request() {
+    // insertion and the call to obtain_permit(), which was verified both by
+    // code inspection and confirmed empirically by running manual tests.
+    if (_server._current_permit.count() > 0) {
+        tlogger.debug("Current service permit is overwritten while its units are still held ({}). "
```
This is better than nothing, and I tend to agree to commit this, but I'm worried that maybe "debug" level is something nobody will ever notice :-(
